src/Materials/materialEvent.ts

```diff
@@ -0,0 +1,346 @@
+import { SmartArray } from "../Misc/smartArray";
+
+declare type Engine = import("../Engines/engine").Engine;
+declare type Scene = import("../scene").Scene;
+declare type BaseTexture = import("./Textures/baseTexture").BaseTexture;
+declare type RenderTargetTexture = import("./Textures/renderTargetTexture").RenderTargetTexture;
+declare type Effect = import("./effect").Effect;
+declare type EffectFallbacks = import("./effectFallbacks").EffectFallbacks;
+declare type MaterialDefines = import("./materialDefines").MaterialDefines;
+declare type UniformBuffer = import("./uniformBuffer").UniformBuffer;
+declare type SubMesh = import("../Meshes/subMesh").SubMesh;
+declare type AbstractMesh = import("../Meshes/abstractMesh").AbstractMesh;
+declare type IAnimatable = import("../Animations/animatable.interface").IAnimatable;
+
+/**
+ * Type for material custom code generators.
+ */
+export type MaterialCustomCodeFunction = (shaderType: string, code: string) => string;
+
+/**
+ * Flags to filter observables in events for material plugins.
+ * @since 5.0.0
+ */
+ export enum MaterialEvent {

```

Comment:
Why do we need a flag type here for the enum ? I do not think a combination makes sense. as the type would be wrong. also there might be a clever way to prevent the ...as... we see everywhere to match an event and a type of data ?

Reply:
Yes, combining flags make no sense but it's a flag because it's how it works at notification level. It's a mask you pass to `notifyObservers` as the second parameter (I kept only the relevant bits):

```typescript
public notifyObservers(eventData: T, mask: number = -1, target?: any, currentTarget?: any, userInfo?: any): boolean {
    ...
    for (var obs of this._observers) {
        if (obs.mask & mask) {
                state.lastReturnValue = obs.callback(eventData, state);
            ...
        }
    }
}
```
As there's a `AND` that is performed between the object mask and passed mask, we must define our events as a bit field.

Reply:
Makes total sense !!!!